crates/subspace-networking/src/behavior/persistent_parameters.rs

```diff
@@ -235,19 +231,28 @@ impl NetworkingParametersRegistry for NetworkingParametersManager {
     async fn remove_known_peer_addresses(&mut self, peer_id: PeerId, addresses: Vec<Multiaddr>) {
         addresses
             .into_iter()
-            .map(|addr| {
-                // remove p2p-protocol suffix if any
-                let mut modified_address = addr.clone();
-
-                if let Some(Protocol::P2p(_)) = modified_address.pop() {
-                    modified_address
-                } else {
-                    addr
-                }
-            })
+            .map(remove_p2p_suffix)
             .for_each(|addr| {
+                // if peer_id is present in the cache
                 if let Some(addresses) = self.known_peers.get_mut(&peer_id) {
-                    addresses.pop(&addr);
+                    // Get mutable reference to last_failed_time for the address without updating
+                    // the item's position in the cache
+                    if let Some(last_failed_time) = addresses.peek_mut(&addr) {
+                        // if we failed previously with this address
+                        if let Some(time) = last_failed_time {
+                            // if we failed less than a day ago
+                            if time.add(chrono::Duration::days(1)) > Utc::now() {
+                                // Update failure time
+                                *last_failed_time = Some(Utc::now())
```

Comment:
I feel like you don't need the whole `crono` crate for this, use `std::time::Instant` in idiomatic way:
```suggestion
                            // if we failed less than a day ago
                            if time.elapsed() < Duration::from_secs(3600 * 24) {
                                // Update failure time
                                last_failed_time.replace(Instant::now());
```

It will also prevent you from getting negative time due to `SystemTime` used by `Utc::now()` being non-monotonic (not sure if underflow will cause crash or ridiculously large time and I don't care), while `Instant` is monotonic and specifically designed for such purposes where you want to measure difference in time.

And lastly I don't think I fully understand why would you want to store last and not first failure time? If you retry over and over again last value will update and not give you any helpful information about what chances are that address is not going to be reachable ever again.

Reply:
We need `chrono` for its `serde` feature which is missing in the `std`. I'm not sure how `Utc::now() being non-monotonic` is affecting us here on the scale of days, please explain. 

The `the last` instead of `the first` is an error (and I even tested it locally !!!) , I refactored the method to a testable alternative and introduced a unit-test.

Reply:
Do you need to serialize timestamps to disk at all BTW? It just feels like the wrong primitive for the job here.

> The the last instead of the first is an error (and I even tested it locally !!!) , I refactored the method to a testable alternative and introduced a unit-test.

I believe you that there could be some kind of error if you change it to the opposite, but it doesn't mean the business logic is correct. Can you describe with words what you wanted to achieve here?